### PR TITLE
Generic Object Definition Actions

### PR DESCRIPTION
**Pull request by jntullo**
Generic Object Definition Actions

Dependent upon https://github.com/ManageIQ/manageiq/pull/15893

@miq-bot add_label enhancement, wip
@miq-bot assign @abellotti 

### General Usage Information
* Can reference a generic object definition by name, id, or href (see "add_attributes" example for how to specify)

#### add_attributes 
```
POST /api/generic_object_definitions
{
  "action": "add_attributes",
  "resources": {
    "attributes": [
    {
      "name": "name",
      "resource": {
        "attributes": {
          "attr1": "string",
          "another": "boolean"
        }
      }
    },
    {
      "id": ":id",
      "resource": {
        "attributes": {
          "attr2": "string"
        }
      }
    },
    {
      "href": "/generic_object_definition/:id",
      "resource": {
        "attributes": {
          "attr3": "string"
        }
      }
    }
  ]}
}
```

```
POST /generic_object_definitions/:id
{
  "action"   : "add_attributes",
  "resource" : {
    "attributes" : {
      "added_attribute1" : "string",
      "added_attribute2" : "boolean"
    }
  }
}
```
#### remove_attributes 
* Accepts same payload as "add_attributes", with different action specified
```
POST /generic_object_definitions/:id
{
  "action"   : "remove_attributes",
  "resource" : {
    "attributes" : {
      "added_attribute1" : "string",
      "added_attribute2" : "boolean"
    }
  }
}
```
### add_methods 
```
POST /generic_object_definitions
{
  "action"    : "add_methods",
  "resources" : [
    { "id" : "id", "resource" : { "methods" : ["method1", "method2"] } }
  ]
}
```
```
POST /generic_object_definitions/:id
{
  "action"   : "add_methods",
  "resource" : {
    "methods" : ["foo", "bar"]
  }
}
```
### remove_methods
* accepts same payload as add_methods with different action specified 
```
POST /generic_object_definitions
{
  "action"    : "remove_methods",
  "resources" : [
    { "id" : "id", "resource" : { "methods" : ["method1", "method2"] } }
  ]
}
```
```
POST /generic_object_definitions/:id
{
  "action"   : "remove_methods",
  "resource" : {
    "methods" : ["foo", "bar"]
  }
}
```
### add_associations
```
POST /generic_object_definitions
{
  "action"    : "add_associations",
  "resources" : [
    { "href" : "/generic_object_definition/:id", "resource" : { "associations" : { "zones" : "AvailabilityZone" } }}
  ]
}
```
```
POST /generic_object_definitions/:id
{
  "action"   : "add_associations",
  "resource" : {
    "associations" : {
      "az"         : "AvailabilityZone",
      "chargeback" : "ChargebackVm"
    }
  }
}
```
### remove_associations 
* accepts same payload as add_associations with different action specified 
```
POST /generic_object_definitions
{
  "action"    : "remove_associations",
  "resources" : [
    { "href" : "/generic_object_definition/:id", "resource" : { "associations" : { "zones" : "AvailabilityZone" } }}
  ]
}
```
```
POST /generic_object_definitions/:id
{
  "action"   : "remove_associations",
  "resource" : {
    "associations" : {
      "az"         : "AvailabilityZone",
      "chargeback" : "ChargebackVm"
    }
  }
}
```

### CODE REVIEW

**Comment by imtayadeway** [comment]
Hi! It looks like your PR has been affected by the recently merged https://github.com/ManageIQ/manageiq-api/pull/40, and as a result will need to be rebased. First, I apologize for any inconvenience caused. After rebasing, you'll need to update some specs in order for them to pass. In particular you'll need to change any path helpers, applying the following pattern:

```
vms_url => api_vms_url
vms_url(vm.id) => api_vm_url(nil, vm) 
vms_url(vm.compressed_id) => api_vm_url(nil, vm.compressed_id)
"#{vms_url(vm.id)}/tags" => api_vm_tags_url(nil, vm)
"#{vms_url(vm.compressed_id}/tags" => api_vm_tags_url(nil, vm.compressed_id)
"#{vms_url(vm.id)}/tags/#{tag.id}" => api_vm_tag_url(nil, vm, tag)
"#{vms_url(vm.compressed_id)}/tags/#{tag.compressed_id}" => api_vm_tag_url(nil, vm.compressed_id, tag.compressed_id)
```

If you run into any issues, please ping me and I will try to help you out.

Many thanks!
Tim

**Comment by abellotti** [comment]
Hi @jntullo maybe add the :forbidden result cases for each of the new action in case they don't have the right role.

File: app/controllers/api/generic_object_definitions_controller.rb [hunk 1]
```diff
@@ -18,13 +17,78 @@ def edit_resource(type, id, data)
...
-      go_def = resource_search(id, type, collection_class(:generic_object_definitions))
+      go_def = fetch_generic_object_definition(type, id, data)
       go_def.destroy!
     rescue => err
       raise BadRequestError, "Failed to delete generic object definition - #{err}"
     end
 
+    def add_attributes_resource(type, id, data)
+      go_def = fetch_generic_object_definition(type, id, data)
+      attributes = data['attributes'] || data['resource']['attributes']
```
Comment (abellotti) [inline app/controllers/api/generic_object_definitions_controller.rb]: isn't this handled by the base class, ie. if data['resource'] exists, that will be the data coming in ?

Comment (jntullo) [reply app/controllers/api/generic_object_definitions_controller.rb]: @abellotti not in the case where the id is the name of the generic object

Comment (abellotti) [reply app/controllers/api/generic_object_definitions_controller.rb]: 👍

**Comment by abellotti** [comment]
Hi @jntullo, looks like https://github.com/ManageIQ/manageiq-api/pull/51 was merged yesterday, can you update the failing tests to use the new get/post/etc. signatures ?  Thanks !!

**Comment by jntullo** [comment]
@abellotti whoops, missed that. this should make them 🍏  😄

**Comment by abellotti** [comment]
Thanks @jntullo for the update.  LGTM!! 👍

**Pull request merged**